### ISSUE TITLE
Recover from connection resets when iterating through playlist

### ISSUE THREAD

**Issue by jcass77** [mcrute/pydora#28]
```iterate_forever``` should probably be able to recover gracefully from exceptions raised due to the Pandora server closing the connection for some reason:

  File "/var/lib/git/pydora/pydora/utils.py", line 120, in iterate_forever
    output = func(*args, **kwargs)
  File "/var/lib/git/pydora/pandora/models/pandora.py", line 24, in get_playlist
    return iter(self._api_client.get_playlist(self.token))
  File "/var/lib/git/pydora/pandora/client.py", line 117, in get_playlist
    includeTrackLength=True))
  File "/var/lib/git/pydora/pandora/client.py", line 88, in __call__
    return self.transport(method, **kwargs)
  File "/var/lib/git/pydora/pandora/transport.py", line 170, in __call__
    result = self._make_http_request(url, data, params)
  File "/var/lib/git/pydora/pandora/transport.py", line 122, in _make_http_request
    r = self._http.post(url, data=data, params=params)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 511, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
    body=body, headers=headers)
  File "/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py", line 376, in _make_request
    httplib_response = conn.getresponse(buffering=True)
  File "/usr/lib/python2.7/httplib.py", line 1034, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 407, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 365, in _read_status
    line = self.fp.readline()
  File "/usr/lib/python2.7/socket.py", line 447, in readline
    data = self._sock.recv(self._rbufsize)
  File "/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 179, in recv
    data = self.connection.recv(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/OpenSSL/SSL.py", line 1320, in recv
    self._raise_ssl_error(self._ssl, result)
  File "/usr/local/lib/python2.7/dist-packages/OpenSSL/SSL.py", line 1178, in _raise_ssl_error
    raise SysCallError(errno, errorcode.get(errno))
SysCallError: (104, 'ECONNRESET')

Perhaps simply retrying the API call up to *n times* with a short delay in between as suggested in https://github.com/kennethreitz/requests/issues/2543 might be an option?

**Comment by jcass77** [comment]
Could be that this is something that should be handled in either urllib3 or requests (also see shazow/urllib3#367)

**Comment by mcrute** [comment]
These types of errors should be handled by the `HTTPAdapter` in [`RetryingSession`](https://github.com/mcrute/pydora/blob/master/pandora/transport.py#L25-L36), or at least that was the theory. I'll try subclassing that to provide better error handling for this case.

**Comment by jcass77** [comment]
I've been hacking on something at https://github.com/jcass77/pydora/blob/709904df881b5d96bf13b1d592d64db1214aedb5/pandora/transport.py#L27-L83

It uses a generic retry decorator to retry the call if the specified exception(s) occur. It is not HTTP session aware at all, so not sure if it will be suitable as a retry handler.

@mcrute is this approach worth considering or would you prefer to handle it differently?

**Comment by mcrute** [comment]
That seems like a pretty good approach. The API is already aware of login timeout exceptions and will retry those so unless there's some other class of error that can't be fixed with a retry this should work.

You may also be able to simplify the logic by using a while loop and decrementing an integer like:

```python
retries_left = 10
while retries_left > 0:
   ... retry ...
   retries_left -= 1
```

**Comment by mcrute** [comment]
Merged #31
